### PR TITLE
merge the sparse hammerblade backend extension to current master code.

### PR DESCRIPTION
**Pull request by zzy82518996**
Now, the last version of sparse hammerblade extension has been merged to current master.
Add more tests in "test_hammerblade_sparse_backend.py"
Un-implemented sparse kernels in native_function.yaml is eliminated, only left backend extension which do not need real implementation of HB sparse kernel.
Including: 
empty_sparse
sparse_dim
dense_dim
_dimI
_dimV
_nnz
is_coalesced
_coalesced_sparse
_indices
indices
_values
values
_copy_sparse_to_sparse

### CODE REVIEW

File: aten/src/ATen/native/sparse/hammerblade/SparseHammerBladeTensor.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+
+#include <ATen/SparseTensorUtils.h>
+//#include <ATen/cuda/CUDAUtils.h>
```
Comment (yodada) [inline aten/src/ATen/native/sparse/hammerblade/SparseHammerBladeTensor.cpp]: Some leftover here ;)

File: c10/core/Backend.h [hunk 2]
```diff
@@ -263,13 +277,15 @@ static inline Backend backendToHammerBlade(Backend b) {
...
+    case Backend::SparseCPU:
+      return Backend::SparseHammerBlade;
     case Backend::CUDA:
     case Backend::HIP:
     case Backend::HammerBlade:
     case Backend::MSNPU:
     case Backend::XLA:
-    case Backend::SparseCPU:
     case Backend::SparseCUDA:
+    case Backend::SparseHammerBlade:
```
Comment (yodada) [inline c10/core/Backend.h]: Hmmm. is this right? it seems we should have SparsehammerBlade to HammerBlade?

File: setup_cosim_build_env.sh [hunk 3]
```diff
@@ -32,7 +32,7 @@ echo "  hb-pytorch lives in $DIR"
 # setup cudalite runtime and pytorch kernel binary paths
 if [ -z "$BRG_BSG_BLADERUNNER_DIR" ]
 then
-  export BSG_MANYCORE_DIR="<path-to-your-cudalite-cosim-runtime>"
+  export BSG_MANYCORE_DIR="/work/shared/users/staff/zz546/bsg_bladerunner/bsg_replicant/libraries"
```
Comment (yodada) [inline setup_cosim_build_env.sh]: Some more leftovers here. ;)

**Review by yodada — COMMENTED** [review]
I think generally it looks good! You might want to clean up the local path you added in setup_cosim. Also, isn't SparseHammerBlade -> HammerBlade supported?

File: c10/core/Backend.h (hunk 2 above)
Comment (zzy82518996) [reply c10/core/Backend.h]: In this way, we need to implement to_dense and to_sparse on HB then, which currently not support. I am not sure whether we need support direct transforming between HB tensor and HB sparse tensor ?

Comment (zzy82518996) [reply c10/core/Backend.h]: So I need to let HammerBlade->HammerBlade and also SparseHammerBlade->Hammerblade right ?

Comment (yodada) [reply c10/core/Backend.h]: I'm not sure either. Maybe we could have a discussion during the meeting today?

Comment (zzy82518996) [reply c10/core/Backend.h]: Sounds good !

**Review by yodada — APPROVED** [review]
Looks good to me! Just one last thing ... I think you created this file accidentally
https://github.com/cornell-brg/hb-pytorch/blob/a2bbf6fd71351a6a1dc0f9cc11ca8f66eeacdfae/c10/core/%2CTensorImpl.h%2C

Could you remove that file? Then be sure to give it a try on cosim, and it's all good to be merged!

**Comment by zzy82518996** [comment]
Thanks Lin, that file is removed, and I test the code again, also pass the cosim.

----- 原始邮件 -----
发件人: "Lin Cheng" <notifications@github.com>
收件人: "cornell-brg/hb-pytorch" <hb-pytorch@noreply.github.com>
抄送: "zzy82158996" <zzy82158996@sjtu.edu.cn>, "Author" <author@noreply.github.com>
发送时间: 星期三, 2020年 3 月 25日 上午 10:48:57
主题: Re: [cornell-brg/hb-pytorch] merge the sparse hammerblade backend extension to current master code. (#20)

@yodada approved this pull request.

Looks good to me! Just one last thing ... I think you created this file accidentally
https://github.com/cornell-brg/hb-pytorch/blob/a2bbf6fd71351a6a1dc0f9cc11ca8f66eeacdfae/c10/core/%2CTensorImpl.h%2C

Could you remove that file? Then be sure to give it a try on cosim, and it's all good to be merged!



-- 
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub:
https://github.com/cornell-brg/hb-pytorch/pull/20#pullrequestreview-381221763

**Pull request merged**